orttraining/tools/scripts/pipeline_model_split.py

```diff
@@ -0,0 +1,376 @@
+import sys
+import onnx
+from onnx import helper
+from onnx import TensorProto
+from onnx import OperatorSetIdProto
+
+# Edge that needs to be cut for the split.
+# If the edge is feeding into more than one nodes, and not all the nodes belong to the same cut,
+# specify those consuming nodes that need to be cut
+class CutEdge:
+    def __init__(self, edgeId, consumingNodes=None):
+        self.edgeId = edgeId
+        self.consumingNodes = consumingNodes
+
+def add_expand_shape(model, name, shape):
+    expand_edge = model.graph.value_info.add()
+    expand_edge.name = name
+    expand_edge.type.CopyFrom(shape)
+
+# Add wait/record/send/recv nodes and split the graph into disconnected subgraphs
+def split_graph(model, edgeIds):
+    ms_domain = 'com.microsoft'
+
+    # Add wait for initial inputs. This needs to be done first before new inputs
+    # are introduced from split
+    initializer_lists = [a.name for a in model.graph.initializer]
+    input_tensors = [
+        value.name for value in model.graph.input if value.name not in initializer_lists]
+
+    input_wait_signal = model.graph.input.add()
+    input_wait_signal.CopyFrom(helper.make_tensor_value_info(
+        'input_wait_signal', onnx.TensorProto.INT64, None))
+
+    input_wait = model.graph.node.add()
+    input_wait.CopyFrom(helper.make_node(
+        'WaitEvent',
+        inputs=['input_wait_signal'],
+        outputs=[],
+        domain=ms_domain))
+
+    for i in input_tensors:
+        for node in model.graph.node:
+            for j in range(len(node.input)):
+                if node.input[j] == i:
+                    print(node.input[j])
+                    node.input[j] = i + '_sync'
+
+    input_wait.input.extend(input_tensors)
+    input_wait.output.extend([i + '_sync' for i in input_tensors])
+
+    # split the graph based on edgeIds
+    upstream_nodes = []
+    output_shapes =[]
+
+    for id in edgeIds:
+        for node in model.graph.node:
+            if len(node.output) >=1 and node.output[0] == id:
```

Comment:
might be easier to build a lookup for output -> node? For big graph, traversal of whole graph multiple times might be expensive.

Reply:
Agree. I am not worried too much here as it is a offline script. But when we move it into ORT, we should consider using a dictionary or some sort for better performance. I'll make a note on this. 

---
In reply to: [400688645](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r400688645) [](ancestors = 400688645)